lib/spack/spack/fetch_strategy.py

```diff
@@ -392,13 +392,30 @@ def __init__(self, *args, **kwargs):
 
     @_needs_stage
     def fetch(self):
-        super(CacheURLFetchStrategy, self).fetch()
+        path = re.sub('^file://', '', self.url)
+
+        # check whether the cache file exists.
+        if not os.path.isfile(path):
+            raise NoCacheError('No cache of %s' % path)
+
+        # remove old symlink if one is there.
+        if os.path.exists(path):
+            os.remove(path)
+
+        # Symlink to local cached archive.
+        self.stage.chdir()
+        basename = os.path.basename(path)
```

Comment:
This chooses the filename in the staging directory based on the cache path. This would look something like "<package-name>-<version>". This may cause problems when the package downloads a single file and does not expand it and has expectations on the name. I'd prefer using self.stage.save_filename instead (which chooses the basename of the default fetcher url when available), although this returns the full path so you might want to use the basename of *that*.